### ISSUE TITLE
FLM Bridge: qwen3_npu::forward() hangs after prefill — NPU sync object never completes

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#210]
## Summary

The FLM bridge engine (npu_engine_flm.cpp) successfully loads models and runs `qwen3_npu::prefill()`, but `qwen3_npu::forward()` hangs indefinitely. The NPU kernel submit succeeds, but the sync object wait never completes.

## Symptoms

```
=== NPU FLM v3 ===
H=1024 NC=28 NV=151936
Init NPU...
Creating qwen3_npu engine...
  Engine created
Loading weights...
  Weights loaded (136ms)
=== Prefill 9 ===
  prefill done, logits size=151936
  465ms (52 ms/tok)
  context length after prefill: 9
  First token: 32313
=== Decode 4 tokens ===
  [0] forward...  <-- HANGS HERE
```

## strace Analysis

The strace shows the NPU kernel submissions succeed but the sync wait never completes:
```
ioctl(3, DRM_IOCTL_AMDXDNA_EXEC_CMD, ...) = 0   // NPU kernel submit OK
ioctl(3, DRM_IOCTL_AMDXDNA_EXEC_CMD, ...) = 0   // NPU kernel submit OK
ioctl(3, DRM_IOCTL_SYNCOBJ_TIMELINE_WAIT, ...) = ? ERESTARTSYS  // NEVER COMPLETES
```

## Attempted Fixes (all failed)

1. Calling `qwen3_npu::set_context_length(9)` before `forward()`
2. Calling `qwen3_npu::checkpoint()` after `prefill()`
3. Calling `qwen3_npu::clear_context()` then re-running
4. Not pre-registering xclbins (letting qwen3_npu handle it)

## Root Cause Suspects

- **Missing AutoModel state**: `qwen3_npu::forward()` may depend on `AutoModel::total_tokens` or other state that only `_shared_insert()` sets up correctly.
- **ABI mismatch**: The `qwen3_npu::Impl` layout when called from our code might differ from when called from FLM CLI.
- **KV cache addressing**: `forward()` might write to an incorrect KV cache position, causing the NPU to hang.

## Evidence

- `flm run qwen3:0.6b` (the CLI binary) works correctly for decode
- `qwen3_npu::prefill()` works correctly
- `qwen3_npu::get_current_context_length()` returns 9 after prefill
- The NPU HW hangs specifically on the instruction sequence that `forward()` generates

## Affected Code

`src/npu_engine_flm.cpp` — FLM bridge engine using `qwen3_npu` directly.

## Workaround

Use the FLM CLI via subprocess wrapper (`npu_engine_flm3.cpp` approach via popen/system) or the REST API server.

Labels: `bug`, `npu`, `flm-bridge`, `hang`

**Comment by bong-water-water-bong** [comment]
## FIXED

The `forward()` hang was caused by missing `addr_qk`, `addr_kv`, `addr_kk`, `addr_l_begin_mha`, `addr_l_end_mha` fields in `LM_Config`. The `qwen3_npu::Impl` reads these configuration values from the model's `config.json` to set up tile allocation. Without them, the Gemm instruction generator produced incorrect DMA sequences that caused the NPU to hang on sync.

**Fix**: Add the correct address fields from `config.json`:
```cpp
lm_cfg.addr_qk = 9216;
lm_cfg.addr_kv = 33792;
lm_cfg.addr_kk = 45056;
lm_cfg.addr_l_begin_mha = 53760;
lm_cfg.addr_l_end_mha = 25088;
```

**Result**: `npu_engine_flm` now runs at **84-97 tok/s** on Qwen3-0.6B using FLM's production xclbins.

Commands:
```
cd ~/npu-infer
./build_and_run.sh [tokens]
```

Build:
```
g++ -std=gnu++17 -O3 -mavx2 -mfma -mavx512f -ffast-math -march=native -fopenmp \
    -I/home/bcloud/FastFlowLM/src/include \
    -I/home/bcloud/fastflowlm-build/src/include \
    -I/home/bcloud/torch2aie/toolchain/xrt/include \
    -c src/npu_engine_flm.cpp -o /tmp/npu_engine_flm.o \
&& g++ /tmp/npu_engine_flm.o -O3 -fopenmp \
    -L/opt/fastflowlm/lib -L/usr/lib/x86_64-linux-gnu \
    -lqwen3_npu -lgemm -lmha -ldequant -llm_head -lq4_npu_eXpress \
    -lxrt_coreutil -lxrt_core -lpthread -ldl -luuid -lrt \
    -o build/npu_engine_flm
```

**Comment by bong-water-water-bong** [comment]
Resolved by architecture refactor. The old `src/npu_engine_flm.cpp` (which directly linked `qwen3_npu` and hung on `forward()`) has been replaced by the `dlopen`-based FlmBridge in `npu-infer/src/flm_bridge.cpp`. The new approach loads FLM's instruction-generation functions dynamically, avoiding the ABI mismatch and missing AutoModel state that caused the hang. The NPU backend now uses this bridge via subprocess worker (backend_npu.cpp) successfully.
